registry/memory/memory.go

```diff
@@ -192,93 +218,199 @@ func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption
-		return nil
-	}
-
-	// refresh TTL and timestamp
-	for _, n := range s.Nodes {
-		if logger.V(logger.DebugLevel, logger.DefaultLogger) {
-			logger.Debugf("Updated registration for service: %s, version: %s", s.Name, s.Version)
+	} else {
+		// refresh TTL and timestamp
+		for _, n := range s.Nodes {
+			if logger.V(logger.DebugLevel, logger.DefaultLogger) {
+				logger.Debugf("Updated registration for service: %s, version: %s", s.Name, s.Version)
+			}
+			srvs[s.Name][s.Version].Nodes[n.Id].TTL = options.TTL
+			srvs[s.Name][s.Version].Nodes[n.Id].LastSeen = time.Now()
 		}
-		m.records[s.Name][s.Version].Nodes[n.Id].TTL = options.TTL
-		m.records[s.Name][s.Version].Nodes[n.Id].LastSeen = time.Now()
 	}
 
+	m.records[options.Domain] = srvs
 	return nil
 }
 
 func (m *Registry) Deregister(s *registry.Service, opts ...registry.DeregisterOption) error {
 	m.Lock()
 	defer m.Unlock()
 
-	if _, ok := m.records[s.Name]; ok {
-		if _, ok := m.records[s.Name][s.Version]; ok {
-			for _, n := range s.Nodes {
-				if _, ok := m.records[s.Name][s.Version].Nodes[n.Id]; ok {
-					if logger.V(logger.DebugLevel, logger.DefaultLogger) {
-						logger.Debugf("Registry removed node from service: %s, version: %s", s.Name, s.Version)
-					}
-					delete(m.records[s.Name][s.Version].Nodes, n.Id)
-				}
-			}
-			if len(m.records[s.Name][s.Version].Nodes) == 0 {
-				delete(m.records[s.Name], s.Version)
-				if logger.V(logger.DebugLevel, logger.DefaultLogger) {
-					logger.Debugf("Registry removed service: %s, version: %s", s.Name, s.Version)
-				}
-			}
-		}
-		if len(m.records[s.Name]) == 0 {
-			delete(m.records, s.Name)
+	// parse the options, fallback to the default domain
+	var options registry.DeregisterOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		options.Domain = defaultDomain
+	}
+
+	// if the domain doesn't exist, there is nothing to deregister
+	services, ok := m.records[options.Domain]
+	if !ok {
+		return nil
+	}
+
+	// if no services with this name and version exist, there is nothing to deregister
+	versions, ok := services[s.Name]
+	if !ok {
+		return nil
+	}
+	version, ok := versions[s.Version]
+	if !ok {
+		return nil
+	}
+
+	// deregister all of the service nodes from this version
+	for _, n := range s.Nodes {
+		if _, ok := version.Nodes[n.Id]; ok {
 			if logger.V(logger.DebugLevel, logger.DefaultLogger) {
-				logger.Debugf("Registry removed service: %s", s.Name)
+				logger.Debugf("Registry removed node from service: %s, version: %s", s.Name, s.Version)
 			}
+			delete(version.Nodes, n.Id)
 		}
+	}
+
+	// if the nodes not empty, we replace the version in the store and exist, the rest of the logic
+	// is cleanup
+	if len(version.Nodes) > 0 {
+		m.records[options.Domain][s.Name][s.Version] = version
+		return nil
+	}
+
+	// if this version was the only version of the service, we can remove the whole service from the
+	// registry and exit
+	if len(versions) == 0 {
```

Comment:
When would `len(versions)` be 0? We haven't deleted this version from `versions` so not sure when this is true.